File: table/src/main/java/com/adsamcik/table/TableAdapter.java
package com.adsamcik.table;

import android.content.Context;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.StyleRes;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.FrameLayout;

import java.util.ArrayList;
import java.util.Collections;

import static com.adsamcik.table.Util.dpToPx;

public class TableAdapter extends BaseAdapter {
	private final ArrayList<Table> tables;
	private final Context context;

	private final int itemMarginPx;

	private final @StyleRes int themeInt;

	public TableAdapter(@NonNull Context context, int itemMarginDp, @StyleRes int themeInt) {
		tables = new ArrayList<>();
		this.context = context.getApplicationContext();
		itemMarginPx = itemMarginDp == 0 ? 0 : dpToPx(context, itemMarginDp);
		this.themeInt = themeInt;
	}

	public void add(Table table) {
		tables.add(table);
	}

	public void clear() {
		tables.clear();
		notifyDataSetChanged();
	}

	public void sort() {
		Collections.sort(tables, (tx, ty) -> tx.appendBehavior - ty.appendBehavior);
		notifyDataSetChanged();
	}

	public void remove(final @AppendBehavior int appendBehavior) {
		if (Build.VERSION.SDK_INT >= 24)
			tables.removeIf(table -> table.appendBehavior == appendBehavior);
		else
			for (int i = 0; i < tables.size(); i++)
				if (tables.get(i).appendBehavior == appendBehavior)
					tables.remove(i--);
		notifyDataSetChanged();
	}

	@Override
	public int getCount() {
		return tables.size();
	}

	@Override
	public Object getItem(int i) {
		return tables.get(i);
	}

	@Override
	public long getItemId(int i) {
		return i;
	}

	@Override
	public View getView(int i, View view, ViewGroup viewGroup) {
		ViewGroup v = (ViewGroup) tables.get(i).getView(context, view, true, themeInt);

		FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) v.getChildAt(0).getLayoutParams();
		lp.setMargins(lp.leftMargin, i > 0 ? itemMarginPx / 2 : itemMarginPx, lp.rightMargin, i < getCount() - 1 ? itemMarginPx / 2 : itemMarginPx);
		v.setLayoutParams(lp);
		return v;
	}
}

File: table/src/main/java/com/adsamcik/table/TableAdapter.kt
package com.adsamcik.table

import android.content.Context
import android.os.Build
import android.support.annotation.StyleRes
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.FrameLayout
import com.adsamcik.table.Util.toPx
import java.util.ArrayList
import kotlin.Any
import kotlin.Comparator
import kotlin.Int
import kotlin.Long

open class TableAdapter(context: Context, itemMarginDp: Int, @param:StyleRes @field:StyleRes
private val themeInt: Int) : BaseAdapter() {
    private val tables: ArrayList<Table> = ArrayList()
    private val context: Context = context.applicationContext

    private val itemMarginPx: Int = itemMarginDp.toPx()

    /**
     * Add table to adapter
     *
     * @param table table
     */
    fun add(table: Table) {
        tables.add(table)
    }

    /**
     * Remove all tables from adapter
     */
    fun clear() {
        tables.clear()
        notifyDataSetChanged()
    }

    /**
     * Sorts tables based on their [AppendBehaviors.AppendBehavior].
     */
    fun sort() {
        tables.sortWith(Comparator { tx, ty -> tx.appendBehavior - ty.appendBehavior })
        notifyDataSetChanged()
    }

    /**
     * Removed all elements with specific [AppendBehaviors.AppendBehavior]
     * @param appendBehavior append behavior
     */
    fun remove(@AppendBehaviors.AppendBehavior appendBehavior: Int) {
        if (Build.VERSION.SDK_INT >= 24)
            tables.removeIf { table -> table.appendBehavior == appendBehavior }
        else {
            var i = 0
            while (i < tables.size) {
                if (tables[i].appendBehavior == appendBehavior)
                    tables.removeAt(i--)
                i++
            }
        }
        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        return tables.size
    }

    override fun getItem(i: Int): Any {
        return tables[i]
    }

    override fun getItemId(i: Int): Long {
        return i.toLong()
    }

    override fun getView(i: Int, view: View?, viewGroup: ViewGroup): View {
        val v = tables[i].getView(context, view, true, themeInt) as ViewGroup

        val lp = v.getChildAt(0).layoutParams as FrameLayout.LayoutParams
        lp.setMargins(lp.leftMargin, if (i > 0) itemMarginPx / 2 else itemMarginPx, lp.rightMargin, if (i < count - 1) itemMarginPx / 2 else itemMarginPx)
        v.layoutParams = lp
        return v
    }
}
